File: benchmarks/fixtures/webapp_go/internal/middleware/ratelimit.go
package middleware

import (
    "sync"

    "webapp_go/pkg/logger"
)

var rlLog = logger.GetLogger("middleware.ratelimit")

// RateLimiter tracks request rates per client.
type RateLimiter struct {
    Requests map[string]int
    Limit    int
    mu       sync.Mutex
}

// NewRateLimiter creates a rate limiter with the specified limit.
func NewRateLimiter(limit int) *RateLimiter {
    rlLog.Info("Creating RateLimiter with limit: %d", limit)
    return &RateLimiter{
        Requests: make(map[string]int),
        Limit:    limit,
    }
}

// RateLimitMiddleware limits request rates per client IP.
func RateLimitMiddleware(limiter *RateLimiter, next Handler) Handler {
    rlLog.Info("Installing rate limit middleware")
    return func(req *Request) *Response {
        ip := req.Headers["X-Forwarded-For"]
        if ip == "" {
            ip = "unknown"
        }
        rlLog.Debug("Rate check for IP: %s", ip)

        limiter.mu.Lock()
        limiter.Requests[ip]++
        count := limiter.Requests[ip]
        limiter.mu.Unlock()

        if count > limiter.Limit {
            rlLog.Warn("Rate limit exceeded for IP: %s (%d requests)", ip, count)
            return &Response{
                Status: 429,
                Body:   map[string]interface{}{"error": "rate limit exceeded"},
            }
        }
        rlLog.Debug("Rate check passed for IP: %s (%d/%d)", ip, count, limiter.Limit)
        return next(req)
    }
}

// Reset clears all rate limit counters.
func (r *RateLimiter) Reset() {
    r.mu.Lock()
    defer r.mu.Unlock()
    rlLog.Info("Resetting rate limiter")
    r.Requests = make(map[string]int)
}

File: benchmarks/fixtures/webapp_php/middleware/rate_limit.php
<?php

namespace App\Middleware;

use function App\Utils\get_logger;

/**
 * Token-bucket rate limiting middleware.
 */
class RateLimitMiddleware
{
    /** @var array<string, int> */
    private array $counters = [];

    /** @param callable $app */
    public function __construct(private $app, private int $limit = 100)
    {
    }

    /**
     * @param array<string, mixed> $request
     *
     * @return mixed
     */
    public function call(array $request)
    {
        $key = $request['ip'] ?? 'anonymous';
        $this->counters[$key] = ($this->counters[$key] ?? 0) + 1;
        if ($this->counters[$key] > $this->limit) {
            get_logger('middleware.rate_limit')->warn("Rate limit exceeded for {$key}");

            return ['error' => 'Too many requests', 'status' => 429];
        }

        return ($this->app)($request);
    }
}
